Plutarch/Api/V1/Value.hs

```diff
@@ -48,3 +51,47 @@ deriving via
   (DerivePConstantViaNewtype Plutus.Value PValue (PMap PCurrencySymbol (PMap PTokenName PInteger)))
   instance
     PConstantDecl Plutus.Value
+
+instance PTryFrom PData (PAsData PTokenName) where
+  type PTryFromExcess PData (PAsData PTokenName) = Flip Term PTokenName
+  ptryFrom' opq = runTermCont $ do
+    (wrapped :: Term _ (PAsData PByteString), unwrapped :: Term _ PByteString) <-
+      tcont $ ptryFrom @(PAsData PByteString) opq
+    tcont $ \f -> pif (pallBS # pisHexDigit # unwrapped) (f ()) (ptraceError "a TokenName must be hex-encoded")
+    tcont $ \f -> pif (plengthBS # unwrapped #<= 32) (f ()) (ptraceError "a TokenName must be at most 32 Bytes long")
+    pure (punsafeFromLifted wrapped, pcon . PTokenName $ unwrapped)
+    where
+      punsafeFromLifted :: PInner a b ~ c => Term s (PAsData c) -> Term s (PAsData a)
+      punsafeFromLifted = punsafeCoerce
+
+      pallBS :: Term s ((PInteger :--> PBool) :--> PByteString :--> PBool)
+      pallBS = phoistAcyclic $
+        plam $ \f str -> plet (plengthBS # str) $ \ln ->
+          let helper :: Term _ (PInteger :--> PBool)
+              helper = pfix #$ plam $ \self i ->
+                pif
+                  (i #< ln)
+                  ( pif
+                      (f #$ pindexBS # str # i)
+                      (self # (i + 1))
+                      (pcon PFalse)
+                  )
+                  (pcon PTrue)
+           in helper # 0
+
+      pisHexDigit :: Term s (PInteger :--> PBool)
+      pisHexDigit = phoistAcyclic $
+        plam $ \chr ->
+          (chr #<= 57 #&& 48 #<= chr)
+            #|| (chr #<= 70 #&& 65 #<= chr)
+            #|| (chr #<= 102 #&& 97 #<= chr)
+
+deriving via
+  DerivePNewtype (PAsData PCurrencySymbol) (PAsData PByteString)
+  instance
+    PTryFrom PData (PAsData PCurrencySymbol)
```

Comment:
It seems that  `CurrencySymbol` should be treated similarly to other hashes, i.e. have fixed-size length of 28 bytes?